base/workerpool.jl

```diff
@@ -51,42 +67,47 @@ function remotecall_pool(rc_f, f, pool::WorkerPool, args...; kwargs...)
         if worker in procs()
             break
         else
-            pool.count = pool.count - 1
+            delete!(pool.workers, worker) # Remove invalid worker from pool
         end
     end
+    return worker
+end
 
+function remotecall_pool(rc_f, f, pool::AbstractWorkerPool, args...; kwargs...)
+    worker = take!(pool)
     try
         rc_f(f, worker, args...; kwargs...)
     finally
-        if worker != 1
-            put!(pool.channel, worker)
-        end
+        # In case of default_worker_pool, the master is implictly considered a worker
+        # till the time new workers are added, and it is not added back to the available pool.
+        # However, it is perfectly valid for other pools to `push!` any worker (including 1)
+        # to the pool. Confirm the same before making a worker available.
+        worker in pool.workers && put!(pool, worker)
     end
 end
 
-
 """
-    remotecall(f, pool::WorkerPool, args...; kwargs...)
+    remotecall(f, pool::AbstractWorkerPool, args...; kwargs...)
 
 Call `f(args...; kwargs...)` on one of the workers in `pool`. Returns a `Future`.
 """
-remotecall(f, pool::WorkerPool, args...; kwargs...) = remotecall_pool(remotecall, f, pool, args...; kwargs...)
+remotecall(f, pool::AbstractWorkerPool, args...; kwargs...) = remotecall_pool(remotecall, f, pool, args...; kwargs...)
 
 
 """
-    remotecall_wait(f, pool::WorkerPool, args...; kwargs...)
+    remotecall_wait(f, pool::AbstractWorkerPool, args...; kwargs...)
```

Comment:
this worker-pool signature of `remotecall_wait` doesn't look like it's present in the rst docs